javascript/computer_science/project_hash_map.md

```diff
@@ -18,17 +18,21 @@ if (index < 0 || index >= buckets.length) {
 
-  1. `hash(key)` takes a key and produces a hash code with it. We did implement a fairly good `hash` function in the previous lesson. You are free to use that, or if you wish, you can conduct your own research. Beware, this is a deep deep rabbit hole.
+  1. `hash(key)` takes a key and produces a hash code with it. We did implement a fairly good `hash` function in the previous lesson, see below. You are free to use that, or if you wish, you can conduct your own research. Beware, this is a deep deep rabbit hole. Also, there is one edge case with long keys that was not taken into consideration in the function or rather how we applied a modulo `%` operator. JavaScript is unable to hold large numbers precisely. At some point, calculations are going to be inacccurate, which significantly increases the chances of collisions. There are a few ways how we could handle it but we recommend that you apply the modulo operator on *each iteration* instead of outside the loop at the end. In that case, we prevent the output from becoming larger than our bucket's length.
 
-      <div class="lesson-note lesson-note--tip" markdown="1">
-        Hash maps could accommodate various data types for keys like numbers, strings, objects. But for this project, only handle keys of type strings.
-      </div>
+  You might find yourself confusing keys with hash codes while accessing key-value pairs later. We would like to stress that the key is what your `hash` function will take as an input. In a way, we could say that the key is important for us only inside the `hash` function. But we never access a bucket directly with the key. Instead we do so with the hash code.
 
-  1. `set(key, value)` takes two arguments, the first is a key and the second is a value that is assigned to this key. If a key already exists, then the old value is overwritten.
+<div class="lesson-note lesson-note--tip" markdown="1">
+  Hash maps could accommodate various data types for keys like numbers, strings, objects. But for this project, only handle keys of type strings.
+</div>
+
+  1. `set(key, value)` takes two arguments, the first is a key and the second is a value that is assigned to this key. If a key already exists, then the old value is overwritten or we can say that we *update* the key's value (e.g. `Carlos` is our key but it is called twice: once with value `I am the old value.`, and once with value `I am the new value.`. From the logic stated above, `Carlos` should contain only the latter value).
+
+  In the meantime, a collision is when *TWO DIFFERENT* keys sit inside the same bucket, because they generate the same hash code (e.g. `Carlos` and `Carla` are both hashed to `3`, so `3` becomes a location for `Carlos` AND `Carla`. However, we know that it is the collision. It means we should find a way how to resolve it — how to *deal with collisions*, which was mentioned in the previous lesson).
```

Comment:
LOVE all this stuff. Great explanations.

The `set()` method here is also a multiline list item, so be sure to get the formatting correct.